### PR TITLE
[wptpr.live] Change PR deployment to use 'pull_request_target'

### LINKED ISSUE #27: wptpr.live - Deployment of PRs from forks
# wprpr.live - Deployment of PRs from forks

Status: In-Review
Authors: smcgruer@chromium.org
Reviewers: @Hexcles, @jugglinmike 
Last Updated: 2020-02-07

## Background

The original design of wptpr.live uses a pair of [GitHub Actions](https://github.com/features/actions) to deploy pull requests to the server. The first action, [pull_request_previews.yml](https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/pull_request_previews.yml), runs periodically and is responsible for determining:

1. which pull requests should be mirrored,
1. creating special refs for those pull requests (which wptpr.live detects and begins mirroring), and
1. creating a [GitHub deployment](https://developer.github.com/v3/guides/delivering-deployments/) for those pull requests.

The second action, [detect_pull_request_preview.yml](https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/detect_pull_request_preview.yml), triggers on GitHub [`deployment` events](https://developer.github.com/v3/activity/events/types/#deploymentevent) and is responsible for monitoring the mirroring on wptpr.live and updating the GitHub pull request with the status and target URL once the deployment is done. Polling [wptpr.live](https://wptpr.live) was chosen as it has to be considered 'untrusted' (it hosts and executes user code), and so cannot be given the necessary GitHub tokens that would be required for it to proactively update GitHub with the mirroring status.

Unfortunately, after deploying this system it was discovered that GitHub was failing to send `deployment` events for deployments created for pull requests that originated from forks. Interestingly it was determined that [GitHub Webhooks](https://developer.github.com/webhooks/) **were** still being sent a `deployment` event in this case. A support ticket to GitHub has so far (~4 business days at time of writing) gone unanswered.

Mirroring pull requests from forks was one of the primary purposes of wptpr.live; the design allows authorized users of wpt to label PRs from external contributors (who usually contribute from forks) as safe to be mirrored. The reviewer can then examine the newly added or modified tests at `wptpr.live/[pr_number]`.

## Proposed Fix - Periodic polling for deployments

The proposed fix is fairly simple; switch the [detect_pull_request_preview.yml](https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/detect_pull_request_preview.yml) script to operate on a periodic basis instead of being triggered from `deployment` events. Instead of being passed information about a specific deployment, the script would fetch the [list of deployments](https://developer.github.com/v3/repos/deployments/#list-deployments) via GitHub APIs, and determine the status of each. It has been verified (via a manual deployment to [this PR](https://github.com/web-platform-tests/wpt/pull/21572)) that deployments created for pull requests from forks do show up in the list.

For each deployment, the script would check the set of associated statuses. If the list is empty or only contains 'pending' status events, then the script would check the status of wptpr.live and update the deployment as appropriate.

One downside to this approach is an increase in calls to the GitHub APIs. This fix adds one request to fetch the list of deployments, and then a call per deployment to fetch the statuses. The system today [already regularly hits API limits](https://github.com/web-platform-tests/wpt.live/issues/26), so those issues will likely have to be addressed before this fix can be made.

### Utilize wpt-pr-bot

As `deployment` events appear to be sent to webhooks without problem, we could utilize the wpt-pr-bot service (which is trusted) to handle deployment status updates. This would involve deleting the [detect_pull_request_preview.yml](https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/detect_pull_request_preview.yml) script and associated code, and migrating the functionality to `wpt-pr-bot`.

This has the upside of not increasing the number of API calls being made. It has the downside of further complicating the wpt-pr-bot codebase and increasing our reliance on a bot that we had actually hoped to remove in favour of GitHub Actions.

### Wait for GitHub to fix their 'bug'

At this point it is unknown whether GitHub consider not sending `deployment` events for pull requests from forks a bug or intended behavior. GitHub support tends to be good once they answer, but their latency can be considerable (O(weeks) is common). We could wait to see if GitHub consider this a bug, at which point we could further wait for them to fix it.

### Disallow usage of wptpr.live for PRs from forks

One obvious option would be to just disallow mirroring to wptpr.live for pull requests from forks. This seems to violate one of the main goals of the project (as noted in the background section), but I note that our original goal was to provide the same service as w3c-test.org, and [this comment](https://github.com/web-platform-tests/wpt.live/issues/2#issuecomment-462499908) appears to imply that w3c-test.org only mirrors patches from trusted persons. Such people are always able to make their pull requests from branches on WPT itself, and do not need to use forks (though some prefer forks).

Given there are reasonable ways to address the problem, this seems like a bad choice, but included for posterity.

### PR DESCRIPTION
**Pull request by stephenmcgruer**
The 'pull_request_target' workflow event is identical to
'pull_request', except it is dispatched in the context of the base
commit rather than the merge commit. This means that only workflows that
already exist in the repository before the PR are run, and using code
that was already in the repository - aka, the code is trustworthy!

Having trustworthy code means we can greatly simplify the design for
deploying code to wptpr.live, and also easily support PRs from forks too -
this change does both.

Fixes https://github.com/web-platform-tests/wpt.live/issues/27

### CODE REVIEW

File: .github/workflows/pull_request_previews.yml
Comment (Hexcles) [inline .github/workflows/pull_request_previews.yml:6]: This is not 100% correct. The threat model here is write access to our GitHub **repo** (including code & settings), not to wptpr.live. wptpr.live by design does not have any sensitive information and the worst thing you could do is spamming (filling up its disk, or using it as free storage), which we guard against by checking author association (and banning abusive users).

Therefore, we are protecting the very powerful **DEPLOY_TOKEN** here *, which is accessible by this workflow (and any script it invokes). We don't want people modifying either and exposing the token, but we can allow members to push whatever they want (from branches or forks) to wptpr.live.

* See the original comment. This token actually has full `repo` access. And I believe it belongs to wpt-pr-bot, which is an admin, so technically you can even push directly to the master branch with this token...

File: tools/ci/pr_preview.py
Comment (Hexcles) [inline tools/ci/pr_preview.py:47]: same

Comment (Hexcles) [inline tools/ci/pr_preview.py:47]: same

File: tools/ci/pr_preview.py
Comment (Hexcles) [inline tools/ci/pr_preview.py:279]: Nit: PEP8 does require two blank lines between top-level functions/classes, but this file only has one empty line everywhere, so either drop the new empty lines or fix the other places for consistency.

File: tools/ci/pr_preview.py
Comment (Hexcles) [inline tools/ci/pr_preview.py:302]: This feels more like debug level to me, but I don't mind leaving it info if you intended to show it by default (debug logs are not).

File: .github/workflows/pull_request_previews.yml
Comment (Hexcles) [inline .github/workflows/pull_request_previews.yml:7]: Nit: maybe rename it to `create-pr-preview` or something? "sync" kind of loses its original meaning with this change as it's no longer a cron job that "synchronizes" states.

File: tools/ci/pr_preview.py
Comment (Hexcles) [inline tools/ci/pr_preview.py:421]: Update this as well.

File: tools/ci/pr_preview.py
Comment (Hexcles) [inline tools/ci/pr_preview.py:406]: Nit
```suggestion
        data = json.load(handle)
```

File: tools/ci/pr_preview.py
Comment (Hexcles) [inline tools/ci/pr_preview.py:363]:
```suggestion
    # No revision to be deployed to wptpr.live
```

**Review by Hexcles — COMMENTED** [review]
Mostly LGTM, still need to review the test

File: .github/workflows/pull_request_previews.yml
Comment (stephenmcgruer) [reply .github/workflows/pull_request_previews.yml:6]: Thank you! That makes a lot more sense to me than what I originally thought the threat model was (I was quite confused as I wrote this comment, because the thread didn't make sense to me...).

Will get back to this comment once we resolve the conversation about `GITHUB_TOKEN` vs `DEPLOY_TOKEN` below...

File: .github/workflows/pull_request_previews.yml
Comment (stephenmcgruer) [reply .github/workflows/pull_request_previews.yml:30]: > as the default GITHUB_TOKEN when running from a fork

I thought the point of `pull_request_target` was that it ran in the context of the base repository not the fork, and thus the `GITHUB_TOKEN` should be read/write?

From https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/ (emphasis mine):

> In order to solve this, we’ve added a new pull_request_target event, which behaves in an almost identical way to the pull_request event with the same set of filters and payload. However, instead of running against the workflow and code from the merge commit, the event runs against the workflow and code from the base of the pull request. This means the workflow is running from a trusted source **and is given access to a read/write token** as well as secrets enabling the maintainer to safely comment on or label a pull request.


To be clear: it is my belief that using `GITHUB_TOKEN` for this workflow should work, and the only reason I am using `DEPLOY_TOKEN` here is for testing, because I am also using `pull_request` for testing (because `pull_request_target` of course wouldn't fire!). I apologise, I referenced this by saying 'the second commit should be reverted' but I should have been more explicit about what the changes in it were.

Comment (Hexcles) [reply .github/workflows/pull_request_previews.yml:30]: Oh sorry I didn't realize that, so yeah using `GITHUB_TOKEN` LGTM.

File: .github/workflows/pull_request_previews.yml
Comment (Hexcles) [reply .github/workflows/pull_request_previews.yml:6]: Yeah, just replace `DEPLOY_TOKEN` with `GITHUB_TOKEN` (and perhaps call out that this token has read/write access regardless of the origin of the PR).

**Comment by foolip** [comment]
Haven't reviewed, but the description is great, hope this all works out!

File: tools/ci/tests/test_pr_preview.py
Comment (Hexcles) [inline tools/ci/tests/test_pr_preview.py:88]: Thank you for fixing this. I think the original implementation abused the fact that an int would be JSON encoded into a number literal, which is a valid git SHA...

File: tools/ci/tests/test_pr_preview.py
Comment (Hexcles) [inline tools/ci/tests/test_pr_preview.py:271]: Does this block still need the temporary repo?

File: tools/ci/tests/test_pr_preview.py
Comment (Hexcles) [inline tools/ci/tests/test_pr_preview.py:290]: Nit
```suggestion
def test_update_mirror_refs_fail_rate_limited():
```

This function no longer uses any API that requires "search" quota.

File: tools/ci/tests/test_pr_preview.py
Comment (Hexcles) [inline tools/ci/tests/test_pr_preview.py:518]:
```suggestion
def test_deploy_fail_rate_limited():
```

**Review by Hexcles — APPROVED** [review]
LGTM % nits

File: .github/workflows/pull_request_previews.yml
Comment (stephenmcgruer) [reply .github/workflows/pull_request_previews.yml]: Done, PTAL at the new version.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/pr_preview.py:299]: I prefer to show it by default, given this is just in CI and knowing what requests fire helps any debugging.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/pr_preview.py]: Done.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/pr_preview.py]: Done.

File: tools/ci/pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/pr_preview.py]: Done.

File: tools/ci/tests/test_pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/tests/test_pr_preview.py:88]: Yep, it was quite a fun one to find and to fix!

And by fun I mean a pain :D.

File: tools/ci/tests/test_pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/tests/test_pr_preview.py]: Good point, I don't believe so. Done.

File: tools/ci/tests/test_pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/tests/test_pr_preview.py]: Good catch! Done.

File: tools/ci/tests/test_pr_preview.py
Comment (stephenmcgruer) [reply tools/ci/tests/test_pr_preview.py]: Done.

**Review by stephenmcgruer — COMMENTED** [review]
I've addressed all the comments (I believe), but am going to wait until next week to merge this to avoid conflicting with the other big change going on (Py3-first switch).

**Review by Hexcles — APPROVED** [review]
Awesome! LGTM.

**Comment by stephenmcgruer** [comment]
Failures on azure-pipelines look unrelated, merging. Let's give this one a spin!

**Pull request merged**
